yarn-project/simulator/src/public/state_manager/state_manager.ts

```diff
@@ -340,27 +340,35 @@ export class PublicPersistableStateManager {
     const exists = instanceWithAddress !== undefined;
 
     const instance = exists ? new SerializableContractInstance(instanceWithAddress) : undefined;
+
+    const isCanonical = contractAddressIsCanonical(contractAddress);
+
+    if (!isCanonical) {
+      // If not canonical, we do the nullifier check. We do this regardless of whether or not the instance exists,
+      // as even for non-existence, we need to generate nullifier check hints.
+
+      // This will internally decide whether to check the nullifier tree or not depending on doMerkleOperations.
+      const nullifierExistsInTree = await this.checkNullifierExists(
+        AztecAddress.fromNumber(CONTRACT_INSTANCE_REGISTRY_CONTRACT_ADDRESS),
+        contractAddress.toField(),
+      );
+      assert(
+        exists == nullifierExistsInTree,
+        `Contract instance for address ${contractAddress} in DB: ${exists} != nullifier tree: ${nullifierExistsInTree}. This is a bug!`,
+      );
+    }
+
```

Comment:
Feels like instead of this coud be re-organised to use the early return instead of this extra if statement. Is the order of checks the following?
1) `isCanonical ? return instance`
2) Nullifier (non)member check
3) Return instance (already checked to undefined if !exists)

We probably don't even need the `if(exists)` branch (the log can be made more generic)